crates/network-monitor/src/lib.rs

```diff
@@ -0,0 +1,241 @@
+
+                    let elapsed = Utc::now() - start_time;
+                    let next_run_in =
+                        (definition_clone.interval as f64 - elapsed.as_seconds_f64()).clamp(1.0, f64::INFINITY);
+                    select! {
+                        _ = cancellation_monitor.cancelled() => { () }
+                        _ = tokio::time::sleep(Duration::from_secs_f64(next_run_in)) => { () }
+                    }
+                }
+            };
+
+            return (
+                (definition.id.clone(), cancellation_token),
+                Box::pin(monitor) as Pin<Box<dyn Future<Output = ()> + Send>>,
+            );
+        })
+        .unzip();
+
+    let mut cancellation_tokens_write = state.cancellation_tokens.lock().await;
+
+    for definition in deleted {
+        cancellation_tokens_write[&definition.id].cancel();
+        cancellation_tokens_write.remove(&definition.id);
+    }
+
+    for (monitor_id, cancellation_token) in new_cancellation_tokens {
+        cancellation_tokens_write.insert(monitor_id, cancellation_token);
+    }
+
+    for monitoring_task in new_monitors {
+        tokio::spawn(monitoring_task);
+    }
+
+    Ok(())
+}
+
+async fn do_ping_monitor(definition: &MonitorDefinition) -> MonitorResult {
+    let start_time = Utc::now();
+
+    let ping_result = async || -> anyhow::Result<TimeDelta> {
+        let runtime = network_scanner_net::runtime::Socket2Runtime::new(None)?;
+        ping::ping_addr(
+            runtime,
+            format!("{hostname}:0", hostname = definition.address),
+            Duration::from_secs(definition.timeout),
+        )
+        .await?;
+        // TODO: send more than 1 ping packet
+
+        Ok(Utc::now() - start_time)
+    }()
+    .await;
+
+    return match ping_result {
+        Ok(time) => MonitorResult {
+            monitor_id: definition.id.clone(),
+            request_start_time: start_time,
+            response_success: true,
+            response_messages: None,
+            response_time: time.as_seconds_f64(),
+        },
+        Err(error) => MonitorResult {
+            // TODO: store error in the result
+            monitor_id: definition.id.clone(),
+            request_start_time: start_time,
+            response_success: false,
+            response_messages: Some(format!("{error:#}").into()),
+            response_time: f64::INFINITY,
+        },
+    };
+}
+
+async fn do_tcpopen_monitor(definition: &MonitorDefinition) -> MonitorResult {
+    MonitorResult {
+        monitor_id: definition.id.clone(),
+        request_start_time: Utc::now(),
+        response_success: false,
+        response_messages: Some("not implemented".into()),
+        response_time: f64::INFINITY,
+    }
+}
+
+pub fn drain_log(state: Arc<State>) -> VecDeque<MonitorResult> {
+    return state.log.drain();
+}
+
+fn get_monitor_config_path(cache_path: &Utf8PathBuf) -> Utf8PathBuf {
+    cache_path.join("monitors_cache.json")
+}
```

Comment:
issue: The current logic assumes that cache_path is a directory and appends a hardcoded filename internally. This creates ambiguity and limits flexibility, as the consumer cannot control the final file path or name. It also introduces hidden behavior that is not obvious to user of the API.

suggestion: This logic should be delegated to the consumer. The library should treat cache_path as the full path to the cache file, allowing consumers to specify both the location and the name.

rationale: Since the library only manages a single cache file, there's no strong reason to impose a fixed internal filename. This change would improve API clarity and make the library more adaptable to various use cases.

Reply:
I agree completely! My original code was reading the config dir path from the existing env variable, but I didn't rethink the implementation when I switched to passing the path through the state struct.